### ISSUE TITLE
[QA] Address Validation QA

### ISSUE THREAD

**Issue by Justin-Pickett** [department-of-veterans-affairs/va.gov-team#3948]
## Background

This epic is to track QA for the address validation functionality that is being added to the VA.gov user profile.

## Tasks 

- [x] Get all scenarios added to test plan/spreadsheet. #5010 (Tze)
- [x] Support QA process (Samara) #5209
- [x] Run QA and create bug tickets. #5012 (Tze)
- [x] Implement bug fixes/updates (Brad)
- [x] Validate bug fixes (Tze) #5012 

## Acceptance Criteria: 

- [x] QA is complete and all bug tickets have been tracked.

**Comment by Justin-Pickett** [comment]
Confirm with QA all use cases are being accounted for and tested for.

**Comment by tlei123** [comment]
@artistgreen / @Samara-Strauss, would you happen to have any centralized documentation of the different user scenarios/flows that generated the mockups in #1688?

Please let me know.  Thanks much!

**Comment by Samara-Strauss** [comment]
@artistgreen was the designer on this project and will be providing you with all flows, documentation, and use cases.

**Comment by artistgreen** [comment]
@tlei123 - yes, I will provide you the scenarios/flows by today.  Thanks!

**Comment by Justin-Pickett** [comment]
@artistgreen and @tzelei have the scenarios/flows been added to the test plan yet? If so, where can I find these or where can I link to them on this ticket? Thanks guys.

**Comment by Samara-Strauss** [comment]
This work needs to be done prior to Tuesday so we can determine if we are actually ready for QA starting on the 18th (day one of sprint 11).

**Comment by artistgreen** [comment]
Thanks @tlei123 - is there anything else you need - outside what was provided in our discussion?
 Let me know if you have any questions about the workflows - I'll keep an eye out and address them quickly.

**Comment by tlei123** [comment]
@Justin-Pickett @Samara-Strauss @artistgreen , I'm working on the plan today.

I'll have a link up by Monday Noon.  :)

**Comment by tlei123** [comment]
Test plan & cases about 85% complete, but I'm putting final completion on hold for now -- some new front-end bugs were opened last week.

Team will determine QA-readiness on Thu 12/19.  If QA-ready, test-cases can be finalized the following day.

**Comment by tlei123** [comment]
@tressaellen @erikphansen , which one of you would be in best position to provide the final design comp link(s)/screenshots after last month's bugfixes?
[Or, just let me know if final design-changes are still in-progress.]

Last month, I was asking @artistgreen about #1688 (Closed Nov. 14), and he was just going to add screenshots here before he said further changes would probably be generated from the then-new bugfixes.

I thought I'd just put this ask on your radars now while we're waiting for bugfix #4278 to get unblocked.

Please let me know.  Thx much!

**Comment by Samara-Strauss** [comment]
@tlei123 while I'm now in a PM role on our team and am no longer a full time designer, I can get you the design files you need. I'll get that to you later today.

**Comment by Samara-Strauss** [comment]
@tzelei this is the flow we're looking at and all it's states. Please let me know what questions you have.

- A user goes to their VA.gov profile and goes to edit their mailing address:

![16  Scenario 3: Edit Address](https://user-images.githubusercontent.com/34068740/72643268-4c59dc80-393c-11ea-81c5-0f23a4a47b7b.png)

- If the address passes address validation on the backend, the modal will close and save the address as expected. If the address does not pass validation, there are **eight possible states** of the address validation screen:

**Address validation determination is overridable; We also have suggestions**

![7  Scenario 1: Override; Suggestions](https://user-images.githubusercontent.com/34068740/72643511-d6a24080-393c-11ea-8878-5d18c40c8aa9.png)

**Address validation determination is overridable; No suggestions**

![13  Scenario 2: Override; No suggestions](https://user-images.githubusercontent.com/34068740/72643522-dc982180-393c-11ea-99d8-6a40220d54e6.png)

**Address validation determination is NOT overridable; We do have suggestions**

*If only one suggestion*

![18  Scenario 3: No override; Only one suggestion](https://user-images.githubusercontent.com/34068740/72643563-f5083c00-393c-11ea-9eb8-6ebcac28fe69.png)

*If more than one suggestion*

![18  Scenario 3: No override; Suggestions](https://user-images.githubusercontent.com/34068740/72643570-fa658680-393c-11ea-9505-7981043f71b0.png)

**Address validation determination is NOT overridable; No suggestions**

![23  Scenario 4: No Override; No Suggestions](https://user-images.githubusercontent.com/34068740/72643589-06514880-393d-11ea-8d96-9bc5ebffee38.png)

**User needs to add a unit number but they CAN override this**

![Add missing unit number with override](https://user-images.githubusercontent.com/34068740/72643595-0b15fc80-393d-11ea-8511-bdcaab0a6ade.png)

**User needs to add a unit number and they CAN NOT override this**

![Add missing unit number without override](https://user-images.githubusercontent.com/34068740/72643601-0e10ed00-393d-11ea-9208-7b1f01601820.png)

**User needs to edit a unit number but they CAN override this**

![Edit bad unit number with override](https://user-images.githubusercontent.com/34068740/72643605-11a47400-393d-11ea-94bf-445ab85a0f7f.png)

**User needs to edit a unit number but they CAN NOT override this**

![Edit bad unit number without override](https://user-images.githubusercontent.com/34068740/72643612-149f6480-393d-11ea-998a-c6bc4b44a4a6.png)

**Comment by tlei123** [comment]
@erikphansen, please provide the specific addresses that will trigger the 8 different validation-states on Staging.

I'll need them to finalize my test-cases.

[IF you actually did demo all 8 states in our Sprint Demos earlier this month, then just let me know and I'll got watch the recording to capture the addresses you used.]

Thanks much!

**Comment by Samara-Strauss** [comment]
@erikphansen — I started [a spreadsheet of addresses for each use case](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/products/identity-personalization/change-address/address-validation/qa/test-cases.md). Feel free to add to it to fill in the blanks.

One thing — do we have any addresses to test the add/edit unit number cases? That's one of the only things I haven't been able to trigger.

**Comment by erikphansen** [comment]
I added a few to the spreadsheet. Also deleted the rows for bad/missing unit numbers with no override as that's not a possible scenario that I've ever seen.

**Comment by tlei123** [comment]
@Samara-Strauss @erikphansen , thx for starting the speardsheet!

I have some follow-up questions:

1. Any way can I help track down the missing addresses?
1. Spreadsheet 1st row (**Override + at least one suggestion**):
  a. Should this row actually be "...one suggestion" (not "...at least one")?  \[I'm assuming this is truly a distinct validation-state from the next row ("...+ two suggestions")]
  b. Assuming this state is indeed distinct ("...one suggestion", not "...at least one"), can we add a matching state-&-screenshot above for this?
1. For the last validation-state **User needs to edit a unit number but they CAN NOT override this**, can we add a matching spreadsheet row?

Pls. let me know.  Thx much!

**Comment by Samara-Strauss** [comment]
> Any way can I help track down the missing addresses?

We've mostly figured this out through trial and error, so there's not much we can do.

> Spreadsheet 1st row (Override + at least one suggestion):
> a. Should this row actually be "...one suggestion" (not "...at least one")? [I'm assuming this is truly a distinct validation-state from the next row ("...+ two suggestions")]
> b. Assuming this state is indeed distinct ("...one suggestion", not "...at least one"), can we add a matching state-&-screenshot above for this?

There could be one or more suggestions, but the design doesn't fundamentally change in this use case (override + suggestions) if there is more than one suggestion — we just list them. In other words, it's only one state to test, whether there is one suggestion or more than one. The only time there is a difference in the designs for there being one or more than one suggestion is for the use case where override is not allowed but there are suggestions (as I've outlined above). Let me know if you have any questions about this.

> For the last validation-state User needs to edit a unit number but they CAN NOT override this, can we add a matching spreadsheet row?

Erik has not been able to find an address to test this case, so he took it out of the spreadsheet.

**Comment by Samara-Strauss** [comment]
> Spreadsheet 1st row (Override + at least one suggestion):
> a. Should this row actually be "...one suggestion" (not "...at least one")? [I'm assuming this is truly a distinct validation-state from the next row ("...+ two suggestions")]
> b. Assuming this state is indeed distinct ("...one suggestion", not "...at least one"), can we add a matching state-&-screenshot above for this?

Just noticed Erik added an address that gives the ability to override + shows two suggestions. This should work for the Override + Suggestion use case. 😃

**Comment by Samara-Strauss** [comment]
@tzelei we have made a change to the address validation functionality so that **override should always be allowed now**. @erikphansen can elaborate on the technical specifics, but this is because we are not getting validation keys for all addresses regardless of whether they are likely bogus or not. 

This means that we actually have fewer use cases to test now for QA. The use cases we are still accounting for are the following. We can further review on our call:

**Address validation determination is overridable; We also have suggestions**

![7  Scenario 1: Override; Suggestions](https://user-images.githubusercontent.com/34068740/72643511-d6a24080-393c-11ea-8878-5d18c40c8aa9.png)

**Address validation determination is overridable; No suggestions**

![13  Scenario 2: Override; No suggestions](https://user-images.githubusercontent.com/34068740/72643522-dc982180-393c-11ea-99d8-6a40220d54e6.png)

**User needs to add a unit number but they CAN override this**

![Add missing unit number with override](https://user-images.githubusercontent.com/34068740/72643595-0b15fc80-393d-11ea-8511-bdcaab0a6ade.png)

**User needs to edit a unit number but they CAN override this**

![Edit bad unit number with override](https://user-images.githubusercontent.com/34068740/72643605-11a47400-393d-11ea-94bf-445ab85a0f7f.png)

**Comment by Samara-Strauss** [comment]
All bugs have been fixed. Closing!
